modules/react-components/src/theme/theme.tsx

```diff
@@ -0,0 +1,354 @@
+/**
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { createContext, Dispatch, useReducer } from "react"; 
+import { Theme, Themes } from "@wso2is/theme";
+
+/*
+ * Themes index less file.
+ */
+const ThemeLess = (theme) => { return `themes-less/themes/${theme}/index.less` };
+
+/*
+ * Theme options type.
+ */
+const themes = Themes as Themes; // Convert Themes array to string literal
+type ThemeTypes = typeof themes[number];
+
+/*
+ * Theme compile options interface.
+ */
+interface ThemeCompileOptionsInterface {
+    "@primaryColor"?: string;
+    "@pageBackground"?: string;
+}
+
+/*
+ * ThemeContext state interface.
+ */
+interface ThemeContextState {
+    appName: string;
+    copyrightText: string;
+    css: string;
+    logo: string;
+    productName: string;
+    styles: ThemeCompileOptionsInterface;
+    theme: ThemeTypes;
+}
+
+/*
+ * ThemeContext interface.
+ */
+interface ThemeContextInterface {
+    dispatch: ({ type }: {type: string}) => void;
+    compile: (options: ThemeCompileOptionsInterface) => void;
+    setAppName: (name: string) => void;
+    setCopyrightText: (text: string) => void;
+    setCSS: (css: string) => void;
+    setLogo: (url: string) => void;
+    setProductName: (name: string) => void;
+    setStyles: (styles: ThemeCompileOptionsInterface) => void;
+    setTheme: (theme: ThemeTypes) => void;
+    state: ThemeContextState;
+}
+
+/**
+ * Initial theme context state.
+ */
+const themeInitialState: ThemeContextState = {
+    appName: "",
+    copyrightText: "",
+    css: "",
+    logo: "",
+    productName: "",
+    styles: {},
+    theme: "default"
+}
+
+/**
+ * ThemeContext reducer actions.
+ */
+const themeContextReducerActions = {
```

Comment:
Would be better to use an enum here IMO. So that we can type cast the reducer actions.